tests/Benchmark/Benchmark.h

```diff
@@ -0,0 +1,142 @@
+﻿//******************************************************************************
+//
+// Copyright (c) Microsoft. All rights reserved.
+//
+// This code is licensed under the MIT License (MIT).
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#pragma once
+
+#include <TestFramework.h>
+#include <type_traits>
+#include <chrono>
+#include <numeric>
+#include <utility>
+#include <memory>
+#include "BenchmarkPublisher.h"
+
+namespace benchmark {
+class BenchmarkCaseBase {
+public:
+    inline void Run();
+
+    // Number of times to run each test, default to 2
+    size_t GetRunCount() const {
+        return 2;
+    }
+
+    // Pre-run and Post-run equivalent to SetUp and TearDown per test case
+    // Setup and Tear down that is performed once per all tests should be done in the constructor/destructor
+    void PreRun() {
+        // Default do nothing
+    }
+
+    void PostRun() {
+        // Default do nothing
+    }
+};
+
+namespace internal {
+
+// Helper method shared between BenchmarkCaseRunner classes
+// This is called to actually run and benchmark the tests
+// Can't be shared internally because parameterized tests need a different inheritance than non
+template <typename T>
+static void __RunCase(T& test, const char* caseName) {
+    using Microseconds = double;
+    size_t runCount = test.GetRunCount();
+    ASSERT_LT(1, runCount);
+    std::vector<Microseconds> results(runCount);
+    @autoreleasepool {
+        for (size_t i = 0; i < runCount; ++i) {
+            test.PreRun();
```

Comment:
That is done in the constructor/destructor of test, this is for things that need to be done once before/after each test run